lib/srv/alpnproxy/local_proxy.go

```diff
@@ -248,9 +250,13 @@ func (l *LocalProxy) StartAWSAccessProxy(ctx context.Context) error {
 }
 
 func (l *LocalProxy) GetCerts() []tls.Certificate {
+	l.certsMu.RLock()
+	defer l.certsMu.RUnlock()
```

Comment:
What prevents the caller from then mutating the certs since it has a copy of the slice? Is this something we should protect against?

Reply:
It's not something we do anywhere, but we could deep copy the slice to prevent that

Reply:
deep copying the slice also requires copying each certificate, do we have any utilities for making generic deep copies?

Alternatively we can remove the exported GetCerts method and instead have a function like `CheckCerts` so that callers from another package can't mutate the slice or its certs within.

Reply:
Perhaps some backstory will clarify how we got here: I added GetCerts to implement a cert checking middleware for local proxy connections. The cert checker was originally in the `lib/srv/alpnproxy` package, but reviewers objected to having `lib/srv/alpnproxy` depend on `lib/client` (it needed a TeleportClient to renew certs). Therefore, the middleware was moved into `lib/client` instead, but I moved the code over without realizing that I could keep the cert checking in the alpnproxy package, somewhat careless I suppose.

I think it's ok to do "unsafe" things within a package, as the scope is limited, but when you export a method it should be safe to use from an outside package's perspective.

Therefore, rather than trying to wrangle deep copying the certs with some reflect hack, I'll just move the cert checking code into the local proxy itself and have the middleware call that function to check if the certs need to be renewed. This turned out to be relatively easy to do. If it were harder, I might have considered just documenting that the caller should not mutate the slice it gets, but that's needlessly fragile.

I'll push after I refactor my integration test.

Reply:
Why does this method even exist anymore? It looks like you're reading `l.cfg.Certs` directly now in most places.

Reply:
The only place I read `l.cfg.Certs` directly is in `CheckDBCerts`, for which I wanted to hold the lock until I'm finished checking the certs. `getCerts` exists as a convenience to get the certs for TLSConfig in `handleDownstreamConnection` and `StartAWSAccessProxy`.